src/main/java/com/google/step/snippet/servlets/FeedbackServlet.java

```diff
@@ -0,0 +1,174 @@
+    if (url == null || !userService.isUserLoggedIn()) {
+      JSONObject json = new JSONObject();
+      json.put(STATUS, NONE);
+      response.setContentType("application/json");
+      response.getWriter().println(json);
+      return;
+    }
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Entity userEntity = getUserEntity(datastore, userService);
+
+    // Get lists of user's upvoted and downvoted cards
+    ArrayList<String> upCards;
+    if (userEntity.getProperty(UPVOTED) == null) {
+      upCards = new ArrayList<String>();
+    } else {
+      upCards = (ArrayList<String>) userEntity.getProperty(UPVOTED);
+    }
+    ArrayList<String> downCards;
+    if (userEntity.getProperty(DOWNVOTED) == null) {
+      downCards = new ArrayList<String>();
+    } else {
+      downCards = (ArrayList<String>) userEntity.getProperty(DOWNVOTED);
+    }
+
+    // Add or remove card from user's list of upvoted and downvoted cards
+    if (request.getParameter(UP) != null) {
+      if (!upCards.contains(url)) {
+        upCards.add(url);
+        if (downCards.contains(url)) {
+          downCards.remove(url);
+        }
+      } else {
+        upCards.remove(url);
+      }
+    } else if (request.getParameter(DOWN) != null) {
+      if (!downCards.contains(url)) {
+        downCards.add(url);
+        if (upCards.contains(url)) {
+          upCards.remove(url);
+        }
+      } else {
+        downCards.remove(url);
+      }
+    }
+    userEntity.setProperty(UPVOTED, upCards);
+    userEntity.setProperty(DOWNVOTED, downCards);
+    datastore.put(userEntity);
+
+    // Retrieve toggle status of upvote and downvote buttons for user
+    String status = NONE;
+    if (userEntity.getProperty(UPVOTED) != null
+        && ((ArrayList<String>) userEntity.getProperty(UPVOTED)).contains(url)) {
+      status = UPVOTED;
+    }
+    if (userEntity.getProperty(DOWNVOTED) != null
+        && ((ArrayList<String>) userEntity.getProperty(DOWNVOTED)).contains(url)) {
+      status = DOWNVOTED;
+    }
+
+    // Calculate the total votes per card
+    long totalUpvotes = 0;
+    long totalDownvotes = 0;
+    Query allUsers = new Query(USER);
+    PreparedQuery results = datastore.prepare(allUsers);
+    for (Entity entity : results.asIterable()) {
+      ArrayList<String> userUpCards = (ArrayList<String>) entity.getProperty(UPVOTED);
+      ArrayList<String> userDownCards = (ArrayList<String>) entity.getProperty(DOWNVOTED);
+      if (userUpCards != null && userUpCards.contains(url)) {
+        totalUpvotes++;
+      }
+      if (userDownCards != null && userDownCards.contains(url)) {
+        totalDownvotes++;
+      }
+    }
+
+    // Update card's feedback entity with new vote total
+    Query.FilterPredicate filter = new Query.FilterPredicate(URL, FilterOperator.EQUAL, url);
+    Query query = new Query(FEEDBACK).setFilter(filter);
+    Entity feedbackEntity = datastore.prepare(query).asSingleEntity();
```

Comment:
The very existence of a Feedback entity seems unnecessary, since you just wrote a bunch of code to determine the number of votes that a piece of feedback has without ever looking at a feedback entity, but again, no time, so okay.